### PR TITLE
Feature/#2 login logout

### PR DESCRIPTION
**Pull request by rufer7**
@dfch please have a look into the code. I'm not yet completely done. The login method of the client is still to be optimized...

### CODE REVIEW

File: src/biz.dfch.CS.Abiquo.Client/Authentication/BasicAuthenticationInformation.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace biz.dfch.CS.Abiquo.Client.Authentication
+{
+    public class BasicAuthenticationInformation : IAuthenticationInformation
+    {
+        private readonly string _username;
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/Authentication/BasicAuthenticationInformation.cs]: @rufer7 I thought we would prefix only static variables with an underscore?

File: src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs [hunk 2]
```diff
@@ -23,8 +25,25 @@
...
+        /// <summary>
+        /// Resets all connection information
+        /// </summary>
+        public void Logout()
+        {
+            this.IsLoggedIn = false;
+
+            this.AuthenticationInformation = null;
+
+            Trace.WriteLine("Logout (Clear/Reset authentication information)");
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs]: @rufer7 this does not really add any useful information except for really logging the invocation. If you only want to log invocation, I would suggest to stick to our approach to log the function name. If you want to enrich the information, add sth like username, tenantid, password (only kidding) or session token.

File: src/biz.dfch.CS.Abiquo.Client/ContractClassForBaseAbiquoClient.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+namespace biz.dfch.CS.Abiquo.Client
+{
+    [ContractClassFor(typeof(BaseAbiquoClient))]
+    abstract class ContractClassForBaseAbiquoClient : BaseAbiquoClient
+    {
+        public override LoginResult Login(string abiquoApiBaseUrl, IAuthenticationInformation authenticationInformation)
+        {
+            Contract.Requires(!string.IsNullOrWhiteSpace(abiquoApiBaseUrl));
+            Contract.Requires(null != authenticationInformation);
+
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/ContractClassForBaseAbiquoClient.cs]: @rufer7 is there no guarantee that the result will be non-null? if no, then add `Contract.Ensures`

File: src/biz.dfch.CS.Abiquo.Client/LoginResult.cs [hunk 4]
```diff
@@ -23,11 +23,11 @@
 
 namespace biz.dfch.CS.Abiquo.Client
 {
-    public enum ELoginResult
+    public enum LoginResult
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/LoginResult.cs]: @rufer7 naming suggestion: `LoginResultEnum`

File: src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs [hunk 5]
```diff
@@ -25,6 +26,17 @@ namespace biz.dfch.CS.Abiquo.Client.v1
 {
     public class AbiquoClient : BaseAbiquoClient
     {
-        
+        public override LoginResult Login(string abiquoApiBaseUrl, Authentication.IAuthenticationInformation authenticationInformation)
+        {
+            Trace.WriteLine(string.Format("START Login (AbiquoApiBaseUrl = '{0}') ...", abiquoApiBaseUrl));
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs]: @rufer7 maybe add the user with which you try to authenticate? and the version number of the client?

File: src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs [hunk 6]
```diff
@@ -23,8 +25,27 @@
...
+        public abstract LoginResultEnum Login(string abiquoApiBaseUrl, IAuthenticationInformation authenticationInformation);
+
+        /// <summary>
+        /// Resets all connection information
+        /// </summary>
+        public void Logout()
+        {
+            this.IsLoggedIn = false;
+
+            this.AuthenticationInformation = null;
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs]: @rufer7 additional thought - maybe we should check if `AuthenticationInformation` is already `null`. then we could log that differently (or not at all).

File: src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs [hunk 7]
```diff
@@ -14,17 +14,48 @@
...
+            DoMakeRequest(Constants.LOGIN_URL_SUFFIX);
+
+            IsLoggedIn = true;
+
+            Trace.WriteLine("END Login SUCCEEDED");
+
+            return LoginResultEnum.Success;
+        }
+
+        private void DoMakeRequest(string urlSuffix)
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs]: @rufer7 do we have the retry mechanism already implemented here? do we want to use the class from the Enterprise Application Building block, or do we reuse the one from CS.Cimi/Appclusive? If we reuse it, I think we should eventually move this into System.Utilities

File: src/biz.dfch.CS.Abiquo.Client/LoginResultEnum.cs [hunk 8]
```diff
@@ -23,11 +23,11 @@
 
 namespace biz.dfch.CS.Abiquo.Client
 {
-    public enum ELoginResult
+    public enum LoginResultEnum
     {
         Success,
```
Comment (dfch) [inline src/biz.dfch.CS.Abiquo.Client/LoginResultEnum.cs]: @rufer7 actually I rethought my suggestion of adding a first enum value. The approach to have the enum value of `0` as the only *good* value is much better and simpler

Comment (rufer7) [inline src/biz.dfch.CS.Abiquo.Client/LoginResultEnum.cs]: ok

File: src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs (hunk 6 above)
Comment (rufer7) [inline src/biz.dfch.CS.Abiquo.Client/BaseAbiquoClient.cs]: @dfch good idea. I'll implement that

File: src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs (hunk 7 above)
Comment (rufer7) [inline src/biz.dfch.CS.Abiquo.Client/v1/AbiquoClient.cs]: @dfch no the retry mechanism is not yet implemented here.

**Comment by rufer7** [comment]
@dfch I now merged the pull request. You can still add comments if desired.

**Pull request merged**